src/lib.rs

```diff
@@ -163,15 +175,25 @@ impl MyNFTCore for NFT {
 }
 
 impl NFT {
-    fn transaction_made(&mut self, transaction_id: u64) -> bool {
+    fn process_transaction(
+        &mut self,
+        transaction_id: u64,
+        action: impl FnOnce(&mut NFT) -> NFTEvent,
+    ) -> NFTEvent {
         let transaction_hash = get_hash(&msg::source(), transaction_id);
-        send_delayed_clear(transaction_hash);
-        if self.transactions.insert(transaction_hash) {
-            false
+
+        if let Some(nft_event) = self.transactions.get(&transaction_hash) {
+            NFTEvent::decode(&mut nft_event.encode().as_slice()).unwrap()
         } else {
-            msg::reply(NFTEvent::TransactionMade, 0)
-                .expect("Error during replying with `NFTEvent::TransactionMade`");
-            true
+            let nft_event = action(self);
+
+            self.transactions.insert(
+                transaction_hash,
+                NFTEvent::decode(&mut nft_event.encode().as_slice()).unwrap(),
```

Comment:
Avoid using `unwrap`. And it's even better no to use `panic`.

Reply:
May be then it is worth to implement `Clone` trait

Reply:
Done.